sharding/nodesSetup.go

```diff
@@ -143,52 +148,67 @@ func (ns *NodesSetup) processMetaChainAssigment() {
 	ns.nrOfMetaChainNodes = 0
 	for id := uint32(0); id < ns.MetaChainMinNodes; id++ {
 		if ns.InitialNodes[id].pubKey != nil {
-			ns.InitialNodes[id].assignedShard = MetachainShardId
+			ns.InitialNodes[id].assignedShard = core.MetachainShardId
+			ns.InitialNodes[id].eligible = true
 			ns.nrOfMetaChainNodes++
 		}
 	}
+
+	hystMeta := uint32(float32(ns.MetaChainMinNodes) * ns.Hysteresis)
+	hystShard := uint32(float32(ns.MinNodesPerShard) * ns.Hysteresis)
+
+	ns.nrOfShards = (ns.nrOfNodes - ns.nrOfMetaChainNodes - hystMeta) / (ns.MinNodesPerShard + hystShard)
 }
 
 func (ns *NodesSetup) processShardAssignment() {
 	// initial implementation - as there is no other info than public key, we allocate first nodes in FIFO order to shards
-	ns.nrOfShards = (ns.nrOfNodes - ns.nrOfMetaChainNodes) / ns.MinNodesPerShard
-
 	currentShard := uint32(0)
 	countSetNodes := ns.nrOfMetaChainNodes
 	for ; currentShard < ns.nrOfShards; currentShard++ {
 		for id := countSetNodes; id < ns.nrOfMetaChainNodes+(currentShard+1)*ns.MinNodesPerShard; id++ {
 			// consider only nodes with valid public key
 			if ns.InitialNodes[id].pubKey != nil {
 				ns.InitialNodes[id].assignedShard = currentShard
+				ns.InitialNodes[id].eligible = true
 				countSetNodes++
 			}
 		}
 	}
 
-	// allocate the rest
+	// allocate the rest to waiting lists
 	currentShard = 0
 	for i := countSetNodes; i < ns.nrOfNodes; i++ {
+		currentShard = (currentShard + 1) % (ns.nrOfShards + 1)
```

Comment:
For these nodes the pubKey is not checked to be different from nil like on line 170l. Should this check be done also here, or this check will be done when creating the new eligible list?

Reply:
right, missed that.
Added the check

Reply:
ok